### PR TITLE
Add self-isolation and change quarantine to be exposure-based

### PR DESCRIPTION
**Pull request by joshwlambert**
This PR addresses #212 and #227. 

The two primary changes in this PR are:

1. adding the ability to self-isolate individuals based on symptoms without requiring a positive test result (#212)
2. changing quarantine to be strictly exposure-based (#227)

## Self-isolation

By allowing the user to specify a proportion of people that self-isolate when symptomatic. These individuals do not get tested and instead isolate themselves a delay after symptom onset. The onset-to-self-isolation delay is separate from the existing onset-to-isolation delay (`onset_to_isolation`) in the model, allowing the user to specify a different (likely shorter) delay to isolation. 

This feature has added a `symptomatic_self_isolate` argument to `event_prob_opts()`, which defaults to 0 for backwards compatibility; and an `onset_to_self_isolation` argument to `delay_opts()` to specify the onset-to-self-isolation delay. These two arguments interact but are specified in different `*_opts()` functions so require cross-checking in the simulation code.

`onset_to_self_isolation` defaults to `NULL` because by default no one self-isolates (and is backwards compatible).

The `cross_check_opts()` function has been added in this PR, and is called in `scenario_sim()` and `outbreak_model()`. It will warn or error if `symptomatic_self_isolate` and `onset_to_self_isolation` are incompatible. Unit tests have been added for `cross_check_opts()`.

Initial cases in `outbreak_setup()` are assumed not to self-isolate. This follows the current assumption that initial cases are missed by contact tracing. 

## Exposure-based quarantine

***Breaking change***

`outbreak_step()` has been restructured so that a case's isolation time is now the earliest of three pathways: 

1. self-isolation
2. a positive test
3. contact tracing. 

Tracing under `quarantine` is now exposure-based, so all traced contacts, including asymptomatic ones, are isolated when their infector is isolated regardless of symptom status (previously asymptomatic cases were never isolated).

`sample_offspring()` reduces post-isolation transmission for asymptomatic cases accordingly. So now asymptomatic individuals do not remain in the same state throughout their infection, but can transition from _asymptomatic_ -> _isolated_ when `quarantine = TRUE`.

---

Unit test expectations for `outbreak_setup()` and `*_opts()` function have been updated, as have the snapshots.

_Not related to the feature but updated in this PR_: argument documentation for `delays`, `event_probs` and `interventions` has been updated to the correct number of output elements.

## Release Notes

* **New Features**
  * Added symptomatic self-isolation functionality allowing users to specify the proportion of symptomatic cases that self-isolate and configure onset-to-isolation delays.
  * Case data output now includes self-isolation status tracking.

* **Improvements**
  * Enhanced validation to prevent incompatible self-isolation parameter combinations and provide clear error guidance.
  * Refined isolation timing logic for self-isolating individuals in transmission simulations.

### CODE REVIEW

File: man/delay_opts.Rd [hunk 1]
```diff
@@ -28,6 +33,19 @@ conditional on \code{gt >= latent_period} (i.e. left-truncated at
...
+the \code{function} accepts a single \code{integer} argument specifying the length
+of the \code{function} output.
+
+By default \code{onset_to_self_isolation} is \code{NULL}. An onset-to-self-isolation
+\code{function} only needs to be specified if a non-zero value is specified to
+\code{symptomatic_self_isolate} in \code{\link[=event_prob_opts]{event_prob_opts()}} (which by default is 0).
+If \code{onset_to_self_isolation} is specified but \code{symptomatic_self_isolate}
+is zero, a warning will be thrown and the \code{onset_to_self_isolation}
+will be ignored; if \code{symptomatic_self_isolate} is non-zero and
+\code{onset_to_self_isolation} is \code{NULL}, \code{\link[=scenario_sim]{scenario_sim()}} will error.}
```
[the remark this answers is not in the document]
Comment (joshwlambert) [reply man/delay_opts.Rd]: Fair point but `scenario_sim()` is seen as the principle user-facing function so I'll leave as is for now. `outbreak_model()` may be moved from exported to internal in a future version of {ringbp}.

File: R/opts.R [hunk 2]
```diff
@@ -73,6 +73,19 @@ offspring_opts <- function(community, isolated, asymptomatic = community) {
...
+#'   the `function` accepts a single `integer` argument specifying the length
+#'   of the `function` output.
+#'
+#'   By default `onset_to_self_isolation` is `NULL`. An onset-to-self-isolation
+#'   `function` only needs to be specified if a non-zero value is specified to
+#'   `symptomatic_self_isolate` in [event_prob_opts()] (which by default is 0).
+#'   If `onset_to_self_isolation` is specified but `symptomatic_self_isolate`
+#'   is zero, a warning will be thrown and the `onset_to_self_isolation`
+#'   will be ignored; if `symptomatic_self_isolate` is non-zero and
+#'   `onset_to_self_isolation` is `NULL`, [scenario_sim()] will error.
```
[the remark this answers is not in the document]
Comment (joshwlambert) [reply R/opts.R]: This is the same suggestion as the comment above, just in a different document. See above for reasoning on not accepting the recommended change.

File: R/outbreak_step.R [hunk 3]
```diff
@@ -122,41 +124,56 @@ outbreak_step <- function(case_data,
...
-      # earliest of symptom onset + delay / infector isolation time
-      default = pmin(ref_time, pmax(onset, infector_isolation_time))
-    )
-  }]
+  # symptomatic, self-isolate, not tested:
+  # isolated at symptom onset + self-isolation delay
+  prob_samples[
+    self_isolate == TRUE,
+    isolated_time := onset + delays$onset_to_self_isolation(.N)
+  ]
```
[the remark this answers is not in the document]
Comment (joshwlambert) [reply R/outbreak_step.R:192]: Fair point. I'm not going to add `cross_check_opts()` to `outbreak_step()` as this will have performance impact. 

Tagging @pearsonca as we've discussed potentially moving some of the exported functions (incl. `outbreak_step()`) to internal and this is another contributing factor to that debate.

[the remark this answers is not in the document]
Comment (pearsonca) [reply R/outbreak_step.R:192]: the generally-easiest approach here is to rip outbreak_step into a public-with-validation outbreak_step and an internal no-validation internal_outbreak_step (or similar naming). i imagine that also means moving some of the validation into the overall public-facing simulation runner previously handled by outbreak_step.

that seem straightforward enough, if now is the time? but maybe approach as a focus PR on that, then rebase this changeset onto that.

[the remark this answers is not in the document]
Comment (joshwlambert) [reply R/outbreak_step.R:192]: I would leave this for a subsequent PR specifically working on the package namespace. 

From my side, the primary (and only?) reason for leaving the functions that `scenario_sim()` calls (`outbreak_model()`, `outbreak_setup()`, `outbreak_step()`) exported is to better enable spin-off projects, as was done in COVID-19. But perhaps the exported/internal status does not influence the ability of projects to extend this model, and it's more to do with the model architecture, I'm genuinely not sure.

It also relates #118.

[the remark this answers is not in the document]
Comment (pearsonca) [reply R/outbreak_step.R:192]: I'd lean towards more to do with the model structure, so agree for another PR whichever way end up pursuing.

an even easier version would be to update the documents for these exported-but-internal-ish functions to make clear that they don't do validation. caveat utor, etc.

File: R/opts.R [hunk 4]
```diff
@@ -83,11 +96,18 @@ offspring_opts <- function(community, isolated, asymptomatic = community) {
 #' )
 delay_opts <- function(incubation_period,
                        onset_to_isolation,
-                       latent_period = 0) {
+                       latent_period = 0,
+                       onset_to_self_isolation = NULL) {
```
Comment (pearsonca) [inline R/opts.R]: Per catch up chat, the one note I have is ambivalence about a NULL default here.

As a rule of thumb, prefer not-NULL arguments (including defaults). This could instead default to a function that returns `Inf` for example.

**Review by pearsonca — APPROVED** [review]
One stylistic-ish note; otherwise LGTM.

File: R/outbreak_step.R [hunk 5]
```diff
@@ -122,41 +124,56 @@ outbreak_step <- function(case_data,
...
+
+  # asymptomatic and symptomatic with false-negative test:
+  # never isolated remain Inf
+  prob_samples[
+    !self_isolate & !asymptomatic & test_positive,
```
Comment (sbfnk) [inline R/outbreak_step.R]: does this imply that if someone is traced and their infector is isolated they still wait until they have symptoms? This seems an odd set up - would make more sense I think to calculate `self_isolation_time` and `traced_isolation_time` separately and then take `pmin(self_isolation_time, traced_isolated_time)`.

This behaviour should also be documented.

File: R/outbreak_step.R [hunk 6]
```diff
@@ -122,41 +124,56 @@ outbreak_step <- function(case_data,
...
   )][,
     # draws a sample to see if this person is asymptomatic
     asymptomatic := runif(.N) < event_probs$asymptomatic
   ][,
     # onset of new case is exposure + incubation period sample
     onset := exposure + delays$incubation_period(.N)
+  ][
+    # draw a sample to see if each person self-isolates if symptomatic
+    asymptomatic == FALSE,
+    self_isolate := runif(.N) < event_probs$symptomatic_self_isolate
```
Comment (sbfnk) [inline R/outbreak_step.R:167]: this means that `asymptomatic == TRUE` have `self_isolate == NA` which seems brittle. Perhaps initiate to `FALSE` above ehre the data.table is constructed?

File: tests/testthat/test-checking.R [hunk 7]
```diff
@@ -38,3 +38,59 @@ test_that("check_dist_func errors when output length does not equal input arg",
...
+    presymptomatic_transmission = 0.5,
+    symptomatic_traced = 0.2,
+    symptomatic_self_isolate = 0.1
+  )
+  expect_error(
+    cross_check_opts(delays = delays, event_probs = event_probs),
+    regexp = paste(
+      "(A non-zero `symptomatic_self_isolate` has been specified)",
+      "(but `onset_to_self_isolation` is `NULL`)",
+      sep = "*"
```
Comment (sbfnk) [inline tests/testthat/test-checking.R]: I assume this is intended as 
```suggestion
      sep = ".*"
```
(i.e. any number of characters)

File: tests/testthat/test-outbreak_step.R [hunk 8]
```diff
@@ -99,8 +99,8 @@ test_that("outbreak_step with > 1 initial infections", {
     interventions = interventions
   )
 
-  # 200 new cases but some are removed from isolation truncating the offspring
-  expect_lt(first_gen_case_data$cases_in_gen, 200)
+  # 200 new cases but some can be removed from isolation truncating the offspring
+  expect_lte(first_gen_case_data$cases_in_gen, 200)
```
Comment (sbfnk) [inline tests/testthat/test-outbreak_step.R:103]: why?

File: R/outbreak_model.R [hunk 9]
```diff
@@ -56,6 +56,7 @@ outbreak_model <- function(initial_cases,
   checkmate::assert_class(event_probs, "ringbp_event_prob_opts")
   checkmate::assert_class(interventions, "ringbp_intervention_opts")
   checkmate::assert_class(sim, "ringbp_sim_opts")
+  cross_check_opts(delays, event_probs)
```
Comment (sbfnk) [inline R/outbreak_model.R:59]: I think this means a warning could be produced in every run in `scenario_sim` - I'd suggest to remove from here (if intended to be an internal function anyway, as you say elsewhere), or use `cli_warn(..., .frequency = "once")` instead (which will lose informativeness with different parameters though).

File: R/outbreak_setup.R
Comment (sbfnk) [inline R/outbreak_setup.R:66]: does this still make sense? Perhaps I'm missing something but to me this reads like symptomatic initial cases get isolated after the onset_to_isolation.

**Review by sbfnk — DISMISSED** [review]
A few general comments:
- Does the model vignette need to be updated with the new model pathway?
- It would be nice to have an integration test for self-isolation, i.e. check that self_isolating cases actually end up with a correct isolated time
- As a general point, this adds another layer of complexity to the model which becomes increasingly harder to parse with ever more forking paths (with opportunities for subtle errors such as the ones I flagged). Not sure I have a solution (and the one I have attempted elsewhere is by no means the final word / only option) but I think it's a concern to bear in mind / try to address.

File: R/opts.R (hunk 4 above)
Comment (joshwlambert) [reply R/opts.R]: Resolved in 3082d98a578197627fdfb23035734f09fa6c3403.

File: R/outbreak_model.R (hunk 9 above)
Comment (joshwlambert) [reply R/outbreak_model.R:59]: Thanks for pointing this out. Resolved in aadc52c3ba238fdc12f41470b67f9e41c8952838 by only running `cross_check_input()` once per `scenario_sim()` or `outbreak_model()` call. I used a base R implementation so didn't require {cli} or {rlang}.

File: R/outbreak_setup.R
Comment (joshwlambert) [reply R/outbreak_setup.R:66]: That's correct, and we intentionally assume that initial cases do not self-isolate. Is there a different way you see this being setup? 

We do not handle test sensitivity in `outbreak_setup()` in case you were thinking about some initial cases not being isolated at all.

File: R/outbreak_step.R (hunk 6 above)
Comment (joshwlambert) [reply R/outbreak_step.R:167]: Resolved in c3e5f4c963c9dd895cd3eb494520e8fd233f14cc.

File: tests/testthat/test-outbreak_step.R (hunk 8 above)
Comment (joshwlambert) [reply tests/testthat/test-outbreak_step.R:103]: This expectation was breaking on the edge case when no cases are removed due to isolation. In most cases (~97%) some cases are removed but in the remaining cases (due to stochastic sampling of generation and isolation times) all 200 cases are kept.

File: R/outbreak_step.R (hunk 5 above)
Comment (joshwlambert) [reply R/outbreak_step.R]: This comment opened pandora's box. 

After thinking about this comment and the logic in `outbreak_step()`, I've opened issue #227 and restructured the isolation logic. This should resolve the isolation time issue in your comment, but also changes the model assumptions around quarantine (asymptomatic cases can now be quarantined). See #227 for details of why I think this change is neccesary.

Ideally this would be in 2 separate PRs, but I've everything in this one. I'll rename the PR and update the PR description to clarify the changes.

File: R/outbreak_step.R [hunk 10]
```diff
@@ -8,15 +8,41 @@
...
+#'
+#' * **Self-isolation**: a symptomatic case self-isolates with probability
+#'   `symptomatic_self_isolate` (from [event_prob_opts()]), entering isolation
+#'   an `onset_to_self_isolation` delay (from [delay_opts()]) after symptom
+#'   onset. Self-isolating cases are not tested.
+#' * **Testing**: a symptomatic case that does not self-isolate is tested and
+#'   returns a positive result with probability `test_sensitivity` (from
+#'   [intervention_opts()]), entering isolation an `onset_to_isolation` delay
+#'   after symptom onset. A false-negative result does not isolate the case
+#'   via this pathway.
```
Comment (pearsonca) [inline R/outbreak_step.R:36]: How do we intend testing to interact with self-isolation? Are they meant to be completely independent, or do we intend something like a false-negative that would happen before self-isolation to preclude self isolation?

Comment (pearsonca) [reply R/outbreak_step.R:36]: the logic below seems to be that *if* an individual would self-isolate, they never consider potentially earlier isolation via testing.

File: R/outbreak_step.R [hunk 11]
```diff
@@ -122,46 +148,83 @@ outbreak_step <- function(case_data,
...
 
   # draw a sample for missing and test result
   prob_samples[
     infector_asymptomatic == FALSE,
     traced := runif(.N) < event_probs$symptomatic_traced
   ][
-    asymptomatic == FALSE,
+    asymptomatic == FALSE & self_isolate == FALSE,
     test_positive := runif(.N) <= interventions$test_sensitivity
   ]
```
Comment (pearsonca) [inline R/outbreak_step.R:177]: here indicates testing conditions on being non-self isolating

**Review by pearsonca — COMMENTED** [review]
I'm not sure we actually get independence of different pathways to outcomes - can double check?

File: R/outbreak_step.R (hunk 10 above)
Comment (joshwlambert) [reply R/outbreak_step.R:36]: My intention is for them to be independent. Someone that wants to self-isolate does not bother testing and someone that does not want to self-isolate requires a test results to enter isolation (both symptom-based). 

Currently, the model does not account for individuals that proactively get tested before showing symptoms, in case this is what you were thinking with your second comment. 

There is a illogicalness to the model if `symptomatic_self_isolate` > 0 and on average `onset_to_self_isolation` is longer than `onset_to_isolation`, because that would mean that people who fall ill and get tested are isolated quicker than people who feel ill and isolate on their own accord. However, this is based on a user's parameterisation and not the model logic, and I guess theoretically this could happen.

File: R/outbreak_step.R (hunk 11 above)
Comment (joshwlambert) [reply R/outbreak_step.R:177]: This is related to my response above, my thinking was that individual who want to self-isolate don't get tested and everyone else requires a test to enter isolation. 

My assumption is that this logic allows realistic scenarios self-isolation for a proportion of the population while maintaining flexibility for the user. However, please let me know if this prevents running a certain scenario you think would be useful.

File: R/outbreak_step.R (hunk 5 above)
Comment (joshwlambert) [reply R/outbreak_step.R]: [TBC]

File: tests/testthat/test-outbreak_step.R (hunk 8 above)
Comment (joshwlambert) [reply tests/testthat/test-outbreak_step.R:103]: [TBC]

File: R/outbreak_setup.R
Comment (joshwlambert) [reply R/outbreak_setup.R:66]: [TBC]

File: R/outbreak_step.R (hunk 10 above)
Comment (joshwlambert) [reply R/outbreak_step.R:36]: [TBC]

File: R/outbreak_step.R (hunk 11 above)
Comment (joshwlambert) [reply R/outbreak_step.R:177]: [TBC]
